### ISSUE TITLE
Support minimal prioritization necessary to enable IETF MoQ base streaming protocol

### ISSUE THREAD

**Issue by wilaw** [w3c/webtransport#431]
There is an ongoing effort within the recently chartered Media Over Quic ([MoQ](https://datatracker.ietf.org/wg/moq/about/)) work group to standardize a simple low-latency media delivery solution for ingest and distribution of media. The use-cases include live streaming, gaming, and media conferencing and these correlate highly with the target [use-cases for WebTransport](https://github.com/w3c/webtransport/blob/main/use-cases.md) within the W3C. As such, it seems reasonable that the initial version of the W3C Webtransport API should at a minimum provide an API surface that enables MoQ base protocol to be implemented in a browser user-agent. 

IETF#115 next week will look consider a [merged proposal between WARP and RUSH](https://www.ietf.org/archive/id/draft-lcurley-warp-02.txt). From the draft text, the requirement around prioritization is:

`This is a strict prioritization scheme, such that any available bandwidth is allocated to streams in ascending priority order.  The sender SHOULD prioritize streams based on the delivery order.  If two streams have the same delivery order, they SHOULD receive equal bandwidth (round-robin).`

Therefore the minimum send prioritization scheme necessary for the web API seems to be one which
1. Allows a strict numeric priority to be set on each WT stream at construction time
2. Allows similar priority to be set against all WT datagrams (as they will be competing with streams for available bandwidth).
3. Requires the user-agent to send the queued high priority data first.  Data with equal priority is round-robined (equal weight). 

This scheme could be extended to incorporate alternative weighting schemes, abstract flows of streams and datagrams, but none of these features are necessary for an initial support of base protocol MoQ and can be deferred to future releases of the API.

Related issues:
Weighted flows of datagrams and short-lived streams [#419](https://github.com/w3c/webtransport/issues/419) 
Datagram vs stream & relative stream prioritization [#62](https://github.com/w3c/webtransport/issues/62)

**Comment by jan-ivar** [comment]
This seems to be what people want first.

As long as the default remains that streams run in parallel with each other, then any minimal ordering mechanism that solves the immediate problem seems fine, as long as it doesn't preclude future iteration. E.g. if we define the following today:
```js
const stream = await wt.createUnidirectionalStream({priority: 123456});
```
...we should be able to expand this later if we want, e.g.:
```js
const stream = await wt.createUnidirectionalStream({priority: 123456, cohort: "A"});
```

s/priority/order/ or whatever.

**Comment by jan-ivar** [comment]
> Allows similar priority to be set against all WT datagrams (as they will be competing with streams for available bandwidth).

Is this needed for MoQ? — I thought what people wanted to solve was inter-ordering of short-lived streams?

**Comment by wilaw** [comment]
Feedback from [IEFT#115 Webtrans meeting](https://notes.ietf.org/notes-ietf-115-webtrans)

`Alan Frindell (AF): MOQ is a baby, lot of debate, can’t wait for MOQ to def priority scheme since that’ll take too long.`

So armed with this feedback, I'd suggestion that we either:
1. Go with no priority scheme at all. The UA will send all datagrams first, followed by a round-robin drain of all established stream queues. Applications can monitor stream queue size and adjust how they write future data. 
2. Implement the most basic priority scheme we can imagine - INt32 numeric priority. Higher numbers have higher priority. The UA drains the send queues in the order of decreasing priority. All streams could have a default priority of 0 and datagrams a priority of MAX INT.  It would be set at stream construction and then also modifiable via property. We could extend this with order, weights or other exotic derivatives in the future once MoQ has matured and/or IETF makes available some future wireline priority scheme which can be leveraged. Any application which doesn't want to enable this priority scheme would simply leave the defaults in place and omit the optional constructor argument.

**Comment by aboba** [comment]
As Luke noted at IETF 115, it is not clear that MoQ requires datagrams, if it supports frame/stream transport. If datagrams aren't supported, priority between datagrams and streams won't be required. 

This leaves priority between streams. In frame/stream transport, it is possible to minimize concurrent sending by using `await writer.close()`.  This ensures that the FIN is written prior to starting to write another frame, effectively prioritizing frames by send order.  If a frame is taking too long to send, `writer.abort()` can be called and the next frame can be sent.

If `writer.close().then()` is used instead, then this allows overlapped writing.   The result is a higher average bitrate, but also much higher frame RTTs and jitter, greater reordering and potentially many more frames aborted due to exceeding the delay threshold.

**Comment by pthatcher** [comment]
I like wilaw's #2.  

The scenario I imagine being important is key frames vs p frames, where you want something like later key frames to be higher priority than earlier ones and you want key frames to be higher priority than the p frames that depend on them.  You could accomplish that with wilaw's #2 with a scheme where you use the even/odd bit to indicate pframe/keyframe and higher numbers to mean later key frames.  I assume similar schemes could be used to divide up priority spaces for similar use cases.

And we can extend this easily to datagrams with something like:

```
const datagrams = transport.sendDatagrams({priority: x});
datagrams.send(d1);
datagrams.send(d2);
datagrams.send(d3);
```

And all these datagrams share the same priority rules/space with streams.

**Comment by jan-ivar** [comment]
> If datagrams aren't supported in MoQ, priority between datagrams and streams isn't likely to be important.

In my mind, the minimal API needed — the one that would be most extensible in the future — would be one that lets JS order a _subset_ of uni streams among themselves only, i.e. order only affects streams participating in ordering:
```js
const frameA = await wt.createUnidirectionalStream({order: 1});
const frameB = await wt.createUnidirectionalStream({order: 2});
const uploadC = await wt.createUnidirectionalStream();
```
I.e. data on `frameB` will be given priority over `frameA`, but equal time with `uploadC` and any datagrams. This seems the most extensible to me. I.e. the default is not `order: 0` but _unordered_.

> it is possible to minimize concurrent sending by using await writer.close(). This ensures that the FIN is written prior to starting to write another frame

This seems extremely costly. My understanding of the goal is not to minimize concurrent sending, but to optimally order concurrent sending without loss of throughput.

**Comment by jan-ivar** [comment]
Meeting:
- Have we narrowed down to one of these?
- "If JS sets a priority (or order whatever), then this means those streams participate in strict ordering, while everything else remains user-agent defined." - This seems like an incremental step from where we are now, and would not preclude extending the API further later.

**Comment by wilaw** [comment]
> > Allows similar priority to be set against all WT datagrams (as they will be competing with streams for available bandwidth).
> 
> Is this needed for MoQ? — I thought what people wanted to solve was inter-ordering of short-lived streams?

Yes - MoQ is looking at a base protocol supporting input drafts from WARP, RUSH and QUICR. While WARP and RUSH use short-lived streams ( of GOPs and frames respectively) , QUICR's default implementation is to use datagrams to deliver fragments of frames.

**Comment by LPardue** [comment]
The suggestion in https://github.com/w3c/webtransport/issues/431#issuecomment-1323753217  is very similar to what I proposed in https://github.com/w3c/webtransport/issues/431#issuecomment-1307999054. For that reason I like it. 

In [extensible priorities](https://www.rfc-editor.org/rfc/rfc9218.html), omitting a parameter means "use the default": urgency = 3, incremental = false (which [means send things serially](https://www.rfc-editor.org/rfc/rfc9218.html#name-incremental)).

The proposal by @jan-ivar is to share available bandwidth across things (i.e. round robin them) in order and then unordered. That's compatible with extensible priorities model if we imagine something like "for WebTransport W3C client API, omitting parameters means the default urgency is X and and incremental is true". 

That way, the proposal like
```
const frameA = await wt.createUnidirectionalStream({order: 1});
const frameB = await wt.createUnidirectionalStream({order: 2});
const uploadC = await wt.createUnidirectionalStream();
```

could easily be extended in the distant future with

```
const frameA = await wt.createUnidirectionalStream({order: 1});
const frameB = await wt.createUnidirectionalStream({order: 2});
const uploadC = await wt.createUnidirectionalStream();
const frameD = await wt.createUnidirectionalStream({order: 1, urgency: 1});
const frameE = await wt.createUnidirectionalStream({order: 2, urgency: 1});
``` 

and that would mean a send schedule like frameE, frameD, frameB, frameA, given equal time share with uploadC and datagrams.

This compatibility approach also means that the model for datagram prioritization and scheduling (that I present at IETF 115 in the MASQUE session https://datatracker.ietf.org/meeting/115/materials/slides-115-masque-datagram-priorities-00) could potentially also fall into alignment in the distant future.

What's missing from the IETF world to help illustrate this compatibility is an `order` extensible priorities parameter, so I'll take it as a task to write that up as an Internet-Draft.
